src/ipc.ts

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright 2022 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { app, autoUpdater, desktopCapturer, ipcMain, powerSaveBlocker } from "electron";
+
+import IpcMainEvent = Electron.IpcMainEvent;
+import { recordSSOSession } from "./protocol";
+import { randomArray } from "./utils";
+import { Settings } from "./settings";
+import { keytar } from "./keytar";
+
+ipcMain.on('setBadgeCount', function(_ev: IpcMainEvent, count: number): void {
+    if (process.platform !== 'win32') {
+        // only set badgeCount on Mac/Linux, the docs say that only those platforms support it but turns out Electron
+        // has some Windows support too, and in some Windows environments this leads to two badges rendering atop
+        // each other. See https://github.com/vector-im/element-web/issues/16942
+        app.badgeCount = count;
+    }
+    if (count === 0) {
+        global.mainWindow?.flashFrame(false);
+    }
+});
+
+let focusHandlerAttached = false;
+ipcMain.on('loudNotification', function(): void {
+    if (process.platform === 'win32' && global.mainWindow && !global.mainWindow.isFocused() && !focusHandlerAttached) {
+        global.mainWindow.flashFrame(true);
+        global.mainWindow.once('focus', () => {
+            global.mainWindow.flashFrame(false);
+            focusHandlerAttached = false;
+        });
+        focusHandlerAttached = true;
+    }
+});
+
+let powerSaveBlockerId: number = null;
+ipcMain.on('app_onAction', function(_ev: IpcMainEvent, payload) {
+    switch (payload.action) {
+        case 'call_state':
```

Comment:
this block is a bit large - might be worth surrounding the contents in `{ }` to make the indentation line up later on